modules/python/python-source.c

```diff
@@ -0,0 +1,505 @@
+                    evt_tag_str("driver", self->super.super.super.id),
+                    evt_tag_str("class", self->class));
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static void
+_py_free_bindings(PythonSourceDriver *self)
+{
+  Py_CLEAR(self->py.class);
+  Py_CLEAR(self->py.instance);
+  Py_CLEAR(self->py.run_method);
+  Py_CLEAR(self->py.request_exit_method);
+  Py_CLEAR(self->py.suspend_method);
+  Py_CLEAR(self->py.wakeup_method);
+}
+
+static gboolean
+_py_init_object(PythonSourceDriver *self)
+{
+  if (!_py_get_attr_or_null(self->py.instance, "init"))
+    {
+      msg_debug("Missing Python method, init()",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("class", self->class));
+      return TRUE;
+    }
+
+  if (!_py_invoke_init(self))
+    {
+      msg_error("Error initializing Python driver object, init() returned FALSE",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("class", self->class));
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static gboolean
+_py_set_parse_options(PythonSourceDriver *self)
+{
+  MsgFormatOptions *parse_options = log_threaded_source_driver_get_parse_options(&self->super.super.super);
+
+  PyObject *py_parse_options = PyCapsule_New(parse_options, NULL, NULL);
+
+  if (!py_parse_options)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating capsule for message parse options",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+      return FALSE;
+    }
+
+  if (PyObject_SetAttrString(self->py.instance, "parse_options", py_parse_options) == -1)
+    {
+      gchar buf[256];
+
+      msg_error("Error setting attribute message parse options",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      Py_DECREF(py_parse_options);
+      return FALSE;
+    }
+
+  Py_DECREF(py_parse_options);
+  return TRUE;
+}
+
+static PyObject *
+py_log_source_post(PyObject *s, PyObject *args, PyObject *kwrds)
+{
+  PyLogSource *self = (PyLogSource *) s;
+  PythonSourceDriver *sd = self->driver;
+
+  PyLogMessage *pymsg;
+
+  static const gchar *kwlist[] = {"msg", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwrds, "O", (gchar **) kwlist, &pymsg))
+    return NULL;
+
+  if (!py_is_log_message((PyObject *) pymsg))
+    {
+      PyErr_Format(PyExc_TypeError, "LogMessage expected in the first parameter");
+      return NULL;
+    }
+
+  /* keep a reference until the PyLogMessage instance is freed */
+  LogMessage *message = log_msg_ref(pymsg->msg);
+  sd->post_message(sd, message);
+
```

Comment:
We shouldn't.

After pushing the message through the pipeline, it will be unrefed and freed by the destination.

The additional ref is required here because of the PyLogMessage instance.

When you create a new PyLogMessage, the LogMessage refcount will be set to 1.
This will be decreased to 0 in the destructor of PyLogMessage, if the message has not been sent out with post_message().

Before posting the message, an extra ref is required, because the corresponding PyLogMessage instance might be stored/cached/etc. on the Python side.

Reply:
Thanks for the explanation. And understand why we had the same ref in the fetcher case too.